File: src/main/java/com/alex323glo/sandbox/rotator/jimpl/Rotator.java
package com.alex323glo.sandbox.rotator.jimpl;

public class Rotator {

    private static final String OPEN_SIGN = "(";
    private static final String CLOSE_SIGN = ")";

    private Rotator() {
    }

    public static String rotateString(String text) {
        if (!text.contains(OPEN_SIGN) || !text.contains(CLOSE_SIGN)) {
            return text;
        }

        int firstIndexOfClose = text.indexOf(CLOSE_SIGN);

        String rightPart = text.substring(firstIndexOfClose + 1);
        String leftAndCentralPart = text.substring(0, firstIndexOfClose);

        int lastIndexOfOpenBeforeClose = leftAndCentralPart.lastIndexOf(OPEN_SIGN);

        String centralPart = leftAndCentralPart.substring(lastIndexOfOpenBeforeClose + 1);
        String leftPart = leftAndCentralPart.substring(0, lastIndexOfOpenBeforeClose);

        String reversedCentralPart = new StringBuilder(centralPart).reverse().toString();
        String newText = leftPart + reversedCentralPart + rightPart;

        return rotateString(newText);
    }
}

File: src/main/kotlin/com/alex323glo/sandbox/rotator/kimpl/procedural/Rotator.kt
package com.alex323glo.sandbox.rotator.kimpl.procedural

private const val OPEN_SIGN = '(';
private const val CLOSE_SIGN = ')';

fun rotateString(text: String): String {
    if (!text.contains(OPEN_SIGN) || !text.contains(CLOSE_SIGN)) {
        return text
    }

    val firstIndexOfClose = text.indexOf(CLOSE_SIGN)

    val rightPart = text.substring(firstIndexOfClose + 1)
    val leftAndCentralPart = text.substring(0 until firstIndexOfClose)

    val lastIndexOfOpenBeforeClose = leftAndCentralPart.lastIndexOf(OPEN_SIGN)

    val centralPart = leftAndCentralPart.substring(lastIndexOfOpenBeforeClose + 1)
    val leftPart = leftAndCentralPart.substring(0 until lastIndexOfOpenBeforeClose)

    val newText = leftPart + centralPart.reversed() + rightPart

    return rotateString(newText)
}